rosetta/static/admin/rosetta/js/rosetta.js

```diff
@@ -1,157 +1,228 @@
-                        }
-                    },
-                    error: function (response) {
-                        a.text(response);
-                    },
+                jsonp(apiUrl, apiData, (response) => {
+                    if (response.code === 200) {
+                        setTranslation(
+                            response.text[0]
+                                .replace(/< ?br>/g, "\n")
+                                .replace(/< ?\/? ?code>/g, "")
+                                .replace(/&lt;/g, "<")
+                                .replace(/&gt;/g, ">"),
+                        );
+                    } else {
+                        setError(response);
+                    }
                 });
             });
         }
     }
 
-    $("td.plural").each(function () {
-        var td = $(this);
-        var trY = parseInt(td.closest("tr").offset().top);
-        $("textarea", $(this).closest("tr")).each(function (j) {
-            var textareaY = parseInt($(this).offset().top) - trY;
-            $($(".part", td).get(j)).css("top", textareaY + "px");
+    // For each translation field textarea
+    document.querySelectorAll(".translation textarea").forEach((textarea, i) => {
+        // Focus on the first textarea on page load
+        if (i === 0) {
+            textarea.focus();
+        }
+
+        // Make textarea heights adapt to their contents on page load
+        textarea.style.height = "auto";
+        textarea.style.height = textarea.scrollHeight + "px";
+
+        // On input
+        textarea.addEventListener("input", function () {
+            // Make textarea heights adapt to their contents
+            this.style.height = "auto";
+            this.style.height = this.scrollHeight + "px";
+
+            // If there are multiple textareas for plurals then align the originals vertically with the textareas
+            alignPlurals();
+
+            // Once users start editing the translation untick the fuzzy checkbox automatically
+            const cb = this.closest("tr").querySelector('td.c input[type="checkbox"]');
+            if (cb.checked) {
+                cb.checked = false;
+            }
+        });
+
+        // On blur show warnings for unmatched variables in translations
+        textarea.addEventListener("blur", function () {
+            const orig = originalForTextarea(this);
+            const variablePattern = /%(?:\([^\s)]*\))?[sdf]|\{[\w\d_]+?\}/g;
+            const origVars = orig.match(variablePattern) || [];
+            const transVars = this.value.match(variablePattern) || [];
+            const everyOrigVarUsed = origVars.every((origVar) => transVars.includes(origVar));
+            const onlyValidVarsUsed = transVars.every((transVar) => origVars.includes(transVar));
+            const valid = everyOrigVarUsed && onlyValidVarsUsed;
+            this.previousElementSibling.classList.toggle("hidden", valid);
         });
```

Comment:
In the testproject there's a string with curly braces: `%{count} observations`. I thought `gettext` works only with the `%(count)d` syntax. I'm not sure how this is supposed to be handled exactly.

Reply:
Yes that looks wrong :thinking: 

This is correct thought:

```python
_("{observations:d} observations").format(observations=10)
```

which will produce
```gettext
#, python-brace-format
msgid "{observations:d} observations"
msgstr ""
```

Reply:
Ah, I didn't realise that's allowed too, I guess there's no reason why it wouldn't work. But the current regex doesn't recognise `{observations:d}` as a variable. We could change it to `\{[^\s}]*\}`, which would follow what the regex does for the percent syntax.

But are the translators always supposed to use the exact same modifier, or is there perhaps a use case to have something different in another language?

(Btw, `compilemessages` won't want about missing variables with the curly braces. It seems to do that only for the percent syntax. I guess this is why all the django docs examples are written that way.)

Reply:
I've changed the regex here: https://github.com/mbi/django-rosetta/pull/295/commits/3fdca8762728bc1e14c090a26a14de97cbfdac40

This expects the exact same `str.format` modifiers to be used in translations.